src/fem/fe_system.cc

```diff
@@ -0,0 +1,416 @@
+}
+
+
+
+template<unsigned int dim, unsigned int spacedim>
+double FESystem<dim,spacedim>::basis_value(const unsigned int i, const arma::vec::fixed<dim> &p) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  return fe_[fe_dof_indices_[dof_basis_[i]].fe_index]->basis_value(fe_dof_indices_[dof_basis_[i]].basis_index, p);
+}
+
+template<unsigned int dim, unsigned int spacedim>
+arma::vec::fixed<dim> FESystem<dim,spacedim>::basis_grad(const unsigned int i, const arma::vec::fixed<dim> &p) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  return fe_[fe_dof_indices_[dof_basis_[i]].fe_index]->basis_grad(fe_dof_indices_[dof_basis_[i]].basis_index, p);
+}
+
+template<unsigned int dim, unsigned int spacedim>
+double FESystem<dim,spacedim>::basis_value_component(const unsigned int i, 
+                                                     const arma::vec::fixed<dim> &p, 
+                                                     const unsigned int comp) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  
+  unsigned int bi = dof_basis_[i];
+  // component index in the base FE
+  int l_comp = comp-fe_dof_indices_[bi].component_offset;
+  OLD_ASSERT(l_comp >= 0 && l_comp < fe_[fe_dof_indices_[bi].fe_index]->n_components(),
+    "Index of component is out of range.");
+
+  if (fe_[fe_dof_indices_[bi].fe_index]->n_components() == 1)
+    return fe_[fe_dof_indices_[bi].fe_index]->basis_value(fe_dof_indices_[bi].basis_index, p);
+
+  return fe_[fe_dof_indices_[bi].fe_index]->basis_value_component(fe_dof_indices_[bi].basis_index, p, l_comp);
+}
+
+template<unsigned int dim, unsigned int spacedim>
+arma::vec::fixed<dim> FESystem<dim,spacedim>::basis_grad_component(const unsigned int i, 
+                                                                   const arma::vec::fixed<dim> &p, 
+                                                                   const unsigned int comp) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  
+  unsigned int bi = dof_basis_[i];
+  // component index in the base FE
+  int l_comp = comp-fe_dof_indices_[bi].component_offset;
+  OLD_ASSERT(l_comp >= 0 && l_comp < fe_[fe_dof_indices_[bi].fe_index]->n_components(),
+    "Index of component is out of range.");
+  
+  if (fe_[fe_dof_indices_[bi].fe_index]->n_components() == 1)
+    return fe_[fe_dof_indices_[bi].fe_index]->basis_grad(fe_dof_indices_[bi].basis_index, p);
+
+  return fe_[fe_dof_indices_[bi].fe_index]->basis_grad_component(fe_dof_indices_[bi].basis_index, p, l_comp);
+}
+
+
+template<unsigned int dim, unsigned int spacedim> inline
+UpdateFlags FESystem<dim,spacedim>::update_each(UpdateFlags flags)
+{
+    UpdateFlags f = flags;
+
+    for (auto fe : fe_)
+      f |= fe->update_each(flags);
+
+    return f;
+}
+
+
+template<unsigned int dim, unsigned int spacedim>
+void FESystem<dim,spacedim>::compute_node_matrix()
+{
```

Comment:
Is this used out of individual finite elements?

Reply:
Postponed to #887

Reply:
This seems simple make it now.

Reply:
Done